src/main/java/Duke.java

```diff
@@ -1,10 +1,287 @@
+                        : "[D] [ ] " + name + " (" + formattedDueDate + ")\n";
+            }
+
+            @Override
+            public void addToFile() {
+                try {
+                    String formattedDueDate = dueDate.format(DateTimeFormatter.ISO_DATE);
+                    FileWriter Writer
+                            = new FileWriter("data/data.txt", true);
+                    Writer.write("D | 0 | " + name + " | " + formattedDueDate + "\n");
+                    Writer.close();
+                } catch (IOException e) {
+                    System.out.println("An error has occurred.");
+                    e.printStackTrace();
+                }
+            }
+        }
+
+        class Events extends Task {
+            LocalDateTime startDateTime;
+            LocalDateTime endDateTime;
+
+            public Events(String name, String startDateTime, String endDateTime) {
+                super(name);
+                this.startDateTime = LocalDateTime.parse(startDateTime, DateTimeFormatter.ofPattern("yyyy-MM-dd HH:mm"));
+                this.endDateTime = LocalDateTime.parse(endDateTime, DateTimeFormatter.ofPattern("yyyy-MM-dd HH:mm"));
+            }
+
+            @Override
+            public String printTask() {
+                DateTimeFormatter formatter = DateTimeFormatter.ofPattern("MMM dd yyyy h:mma");
+                String formattedStartDateTime = startDateTime.format(formatter);
+                String formattedEndDateTime = endDateTime.format(formatter);
+
+                return (isDone) ? "[E] [X] " + name + " (" + formattedStartDateTime + " - " + formattedEndDateTime + ")\n"
+                        : "[E] [ ] " + name + " (from " + formattedStartDateTime + " to " + formattedEndDateTime + ")\n";
+            }
+
+            @Override
+            public void addToFile() {
+                try {
+                    FileWriter writer = new FileWriter("data/data.txt", true);
+                    DateTimeFormatter fileFormatter = DateTimeFormatter.ofPattern("yyyy-MM-dd HH:mm");
+                    String formattedStartDateTime = startDateTime.format(fileFormatter);
+                    String formattedEndDateTime = endDateTime.format(fileFormatter);
+
+                    writer.write("E | " + (isDone ? "X" : " ") + " | " + name + " | " + formattedStartDateTime + " | " + formattedEndDateTime + "\n");
+                    writer.close();
+                } catch (IOException e) {
+                    System.out.println("An error has occurred.");
+                    e.printStackTrace();
+                }
+            }
+        }
+
+        Scanner scanner = new Scanner(System.in);
+        ArrayList<Task> tasks = new ArrayList<>();
+        int count = 0;
+        File file = new File("data/data.txt");
+
+        try {
+            Scanner fileReader = new Scanner(file);
+            while (fileReader.hasNextLine()) {
```

Comment:
I can see that you are loading the file into the array. However, would it be better to abstract the logic somewhere else?

Reply:
What if one day maybe the input system change and you are reading from a  database instead?

Or maybe the input is of a different format (e.g. JSON, XML, YAML)